<!-- -->
<launch>

  <!-- Configurations -->
  <arg name="number_of_accumulations" default="5" />
  <arg name="topic_pointcloud_ref" default="/points_raw" />
  <arg name="topic_pointcloud_ground" default="/points_ground" />
  <arg name="output_frame_id" default="/velodyne" />
  
  <!-- node for ref -->
  <group ns="ref">
    <node pkg="lidar_fast_and_furious" type="points_accumulator_node" name="points_accumulator" output="screen">
      <param name="number_of_accumulations" value="$(arg number_of_accumulations)" />
      <param name="topic_pointcloud" value="$(arg topic_pointcloud_ref)" />
      <param name="use_current_velocity" value="false" />
      <param name="remove_points_negative_z" value="false" />
      <param name="output_frame_id" value="$(arg output_frame_id)" />
    </node>
  </group>

  <!-- node for ground -->
  <group ns="ground">
    <node pkg="lidar_fast_and_furious" type="points_accumulator_node" name="points_accumulator" output="screen">
      <param name="number_of_accumulations" value="$(arg number_of_accumulations)" />
      <param name="topic_pointcloud" value="$(arg topic_pointcloud_ground)" />
      <param name="use_current_velocity" value="true" />
      <param name="remove_points_negative_z" value="false" />
      <param name="topic_current_velocity" value="/ref/estimated_velocity" />
      <param name="output_frame_id" value="$(arg output_frame_id)" />
    </node>
  </group>

  <!-- Configurations for ray_ground_filter -->
  <arg name="sensor_height" default="1.0" />
  <arg name="clipping_height" default="0.2" />
  <arg name="min_point_distance" default="1.85" />
  <arg name="radial_divider_angle" default="0.08" />
  <arg name="concentric_divider_distance" default="0.01" />
  <arg name="local_max_slope" default="8" />
  <arg name="general_max_slope" default="5" />
  <arg name="min_height_threshold" default="0.05" />
  <arg name="reclass_distance_threshold" default="0.2" />
  <arg name="no_ground_point_topic" default="/points_no_ground" />
  <arg name="ground_point_topic" default="/points_ground" />
  
  <!-- rosrun points_preprocessor ray_ground_filter -->
  <node pkg="points_preprocessor" type="ray_ground_filter" name="ray_ground_filter" output="screen">
    <param name="input_point_topic" value="$(arg topic_pointcloud_ref)" />
    <param name="sensor_height" value="$(arg sensor_height)" />
    <param name="clipping_height" value="$(arg clipping_height)" />
    <param name="min_point_distance" value="$(arg min_point_distance)" />
    <param name="radial_divider_angle" value="$(arg radial_divider_angle)" />
    <param name="concentric_divider_distance" value="$(arg concentric_divider_distance)" />
    <param name="local_max_slope" value="$(arg local_max_slope)" />
    <param name="general_max_slope" value="$(arg general_max_slope)" />
    <param name="min_height_threshold" value="$(arg min_height_threshold)" />
    <param name="reclass_distance_threshold" value="$(arg reclass_distance_threshold)" />
    <param name="no_ground_point_topic" value="$(arg no_ground_point_topic)" />
    <param name="ground_point_topic" value="$(arg ground_point_topic)" />
  </node>
  
</launch>
